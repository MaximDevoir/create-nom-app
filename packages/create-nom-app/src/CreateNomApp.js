import { sync as spawnSync } from 'cross-spawn'

import path from 'path'
import fs from 'fs-extra'

import pkg from '../package'

const defaultOptions = {
  projectDirectory: null,
  packageManager: null,
  gitBinary: null
}

class CreateNomApp {
  constructor(projectName, options) {
    // TODO: Do recursive deep-assign
    this.projectName = projectName
    this.options = Object.assign({}, defaultOptions, options)
  }

  ensureProjectDir() {
    fs.ensureDirSync(this.options.projectDirectory)
  }

  installPackages() {
    console.log('package manager is', this.options.packageManager.manager, 'at', this.options.packageManager.binary)
    const cnaMajorVersion = pkg.version.split('.')[0]
    const dependencies = [`nom-scripts@^${cnaMajorVersion}`]
    let installCommand

    // eslint-disable-next-line default-case
    switch (this.options.packageManager.manager) {
      case 'npm':
        installCommand = ['install', '--save', '--verbose']
        break
      case 'yarn':
        installCommand = ['add']
        break
    }

    installCommand.push(...dependencies)
    const install = spawnSync(this.options.packageManager.binary, installCommand, {
      cwd: this.options.projectDirectory,
      stdio: 'inherit'
    })

    if (install.status !== 0) {
      console.error(`Failed to execute ${this.options.packageManager.binary} ${installCommand.join(' ')}`)
      process.exit(1)
    }
  }

  handoff() {
    const nomScriptsBin = path.resolve(this.options.projectDirectory,
      path.join(this.options.projectDirectory, 'node_modules', '.bin', 'nom-scripts'))
    const handoffOptions = JSON.stringify({
      projectName: this.projectName,
      projectDirectory: this.options.projectDirectory,
      packageManager: this.options.packageManager,
      gitBinary: this.options.gitBinary
    })
    const handoffProc = spawnSync(nomScriptsBin, ['init', '--options-string', handoffOptions], {
      stdio: 'inherit',
      cwd: this.options.projectDirectory
    })

    if (handoffProc.status !== 0) {
      console.log('Create Nom App failed during handoff to local create-nom-app project.')
      process.exit(handoffProc.status)
    }
  }
}

export default CreateNomApp
