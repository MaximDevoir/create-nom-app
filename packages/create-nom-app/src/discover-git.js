import { sync as spawnSync } from 'cross-spawn'
import { sync as whichSync } from 'which'

// Semver regex string provided by https://github.com/semver/semver/issues/232#issue-48635632
const semverRegex = /(0|[5-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(-(0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(\.(0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*)?(\+[0-9a-zA-Z-]+(\.[0-9a-zA-Z-]+)*)?/g

const gitInfo = {
  gitInstalled: null,
  gitVersion: null,
  gitBinary: null
}

function discoverGit() {
  const binary = whichSync('git', { nothrow: true })

  if (binary === null) {
    gitInfo.gitInstalled = false
    return
  }

  const version = spawnSync(binary, ['--version'], { stdio: 'pipe' }).stdout.toString().trim()

  gitInfo.gitInstalled = true
  gitInfo.gitVersion = version.match(semverRegex)?.[0] || ''
  gitInfo.gitBinary = binary
}

export default {
  discoverGit,
  gitInfo
}
