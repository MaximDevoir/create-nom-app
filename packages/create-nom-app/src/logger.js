// TODO: This needs to go in its own package
import winston from 'winston'

/**
 * Uses NPM levels for logging:
 *  error: 0
 *  warn: 1
 *  info: 2
 *  verbose: 3
 *  debug: 4
 *  silly: 5
 */
const logger = winston.createLogger({
  format: winston.format.simple(),
  transports: [
    new winston.transports.Console({
      level: (process.env.LOG_LEVEL || 'info').toLowerCase(),
      handleExceptions: true,
      json: false,
      colorize: true
    })
  ],
  exitOnError: false
})

export default logger
