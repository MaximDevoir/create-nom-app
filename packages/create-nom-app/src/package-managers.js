import { execSync } from 'child_process'
import { sync as whichSync } from 'which'

const packageManagers = {}
// TODO: Bring this module to its own package, preferably using
// nom-scripts/create-nom-app to handle dependencies.

/**
 * Searches for a `manager` on the system and, if it exists, adds information
 * about the manager to `packageManagers`.
 *
 * @param {string}  manager Name of the package manager
 * @param {boolean} ignoreCache Whether or not to ignore checking the cache.
 * @returns {undefined} Always returns `undefined`.
 */
function discoverManager(manager, ignoreCache = false) {
  // The `manager` was previously discovered and is in the cache.
  if (ignoreCache === false
     && Object.prototype.hasOwnProperty.call(packageManagers, manager)
  ) {
    return
  }

  const binary = whichSync(manager, { nothrow: true })

  if (binary === null) {
    return
  }

  const version = execSync(`${binary} --version`, { stdio: 'pipe' }).toString().trim()

  packageManagers[manager] = {
    binary,
    version
  }
}

/**
 * Search for an array of managers.
 *
 * @param {string|array} searchFor A list of managers to search for.
 * @param {boolean|undefined} ignoreCache See discoverManager
 */
function discoverMany(managers = [], ignoreCache) {
  const searchFor = []

  if (typeof managers === 'string') {
    searchFor.push(managers)
  } else {
    searchFor.push(...managers)
  }

  searchFor.forEach((manager) => {
    discoverManager(manager, ignoreCache)
  })
}

/**
 * Discover common package managers: npm, yarn.
 */
function discoverCommon() {
  discoverMany(['npm', 'yarn'])
}

/**
 * Clears the cache of package managers.
 */
function clearCache() {
  // eslint-disable-next-line no-restricted-syntax
  for (const manager in packageManagers) {
    if (Object.prototype.hasOwnProperty.call(packageManagers, manager)) {
      delete packageManagers[manager]
    }
  }
}

/**
 * Check if a `manager` exists in `packageManagers`.
 *
 * @param {string} manager The manager to check.
 * @returns {boolean} Whether or not `manager` is in `packageMangers`
 */
function hasManager(manager) {
  if (Object.prototype.hasOwnProperty.call(packageManagers, manager)) {
    return true
  }

  return false
}

/**
 * Get a `manager` from `packageManagers`. If the manager can not be found, an
 * empty object is returned.
 *
 * Use`.hasManager()` to check if a manager exists in `packageManagers`.
 *
 * @param {string} manager The name of the package manager to get.
 * @returns {object}
 */
function getManager(manager) {
  if (hasManager(manager)) {
    return packageManagers[manager]
  }

  return {}
}

export default {
  hasManager,
  getManager,
  discoverCommon,
  discoverMany,
  discoverManager,
  packageManagers,
  clearCache
}
