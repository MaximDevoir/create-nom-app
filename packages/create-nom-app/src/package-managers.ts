import { sync as spawnSync } from 'cross-spawn'
import { sync as whichSync } from 'which'

interface PackageManager {
  binary: string
  version: string
}

const packageManagers: {
  [key: string]: PackageManager
} = {}
// TODO: Bring this module to its own package, preferably using
// nom-scripts/create-nom-app to handle dependencies.

/**
 * Searches for a `manager` on the system and, if it exists, adds information
 * about the manager to `packageManagers`.
 *
 * @param {string}  manager Name of the package manager
 * @param {boolean} ignoreCache Whether or not to ignore checking the cache.
 * @returns {undefined} Always returns `undefined`.
 */
function discoverManager(manager: string, ignoreCache = false): void {
  // The `manager` was previously discovered and is in the cache.
  if (ignoreCache === false && Object.prototype.hasOwnProperty.call(packageManagers, manager)) {
    return
  }

  const binary = whichSync(manager, { nothrow: true })

  if (binary === null) {
    return
  }

  const version = spawnSync(binary, ['--version'], {
    stdio: 'pipe'
  })
    .stdout.toString()
    .trim()

  packageManagers[manager] = {
    binary,
    version
  }
}

/**
 * Search for an array of managers.
 *
 * @param {string|array} searchFor A list of managers to search for.
 * @param {boolean|undefined} ignoreCache See discoverManager
 */
function discoverMany(managers: string[] = [], ignoreCache?: boolean): void {
  const searchFor: string[] = []

  if (typeof managers === 'string') {
    searchFor.push(managers)
  } else {
    searchFor.push(...managers)
  }

  searchFor.forEach(manager => {
    discoverManager(manager, ignoreCache)
  })
}

/**
 * Discover common package managers: npm, yarn.
 */
function discoverCommon(): void {
  discoverMany(['npm', 'yarn'])
}

/**
 * Clears the cache of package managers.
 */
function clearCache(): void {
  // eslint-disable-next-line no-restricted-syntax
  for (const manager in packageManagers) {
    if (Object.prototype.hasOwnProperty.call(packageManagers, manager)) {
      delete packageManagers[manager]
    }
  }
}

/**
 * Check if a `manager` exists in `packageManagers`.
 *
 * @param {string} manager The manager to check.
 * @returns {boolean} Whether or not `manager` is in `packageMangers`
 */
function hasManager(manager: string): boolean {
  if (manager in packageManagers) {
    return true
  }

  return false
}

/**
 * Get a `manager` from `packageManagers`. If the manager can not be found, an
 * empty object is returned.
 *
 * Use`.hasManager()` to check if a manager exists in `packageManagers`.
 *
 * @param {string} manager The name of the package manager to get.
 * @returns {object}
 */
function getManager(manager: string): PackageManager {
  if (hasManager(manager)) {
    return packageManagers[manager]
  }

  return {
    version: '',
    binary: ''
  }
}

export default {
  hasManager,
  getManager,
  discoverCommon,
  discoverMany,
  discoverManager,
  packageManagers,
  clearCache
}
