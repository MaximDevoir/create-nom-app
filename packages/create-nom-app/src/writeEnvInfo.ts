// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
// @ts-ignore
import envinfo from 'envinfo'
import chalk from 'chalk'

// eslint-disable-next-line @typescript-eslint/no-var-requires
const { version } = require('../package.json')

async function writeEnvInfo(): Promise<void> {
  console.log(chalk.bold('\nEnvironment Info:'))
  process.stdout.write(`  Create Nom App: ${version}`)
  await envinfo
    .run(
      {
        System: ['OS', 'CPU'],
        Binaries: ['Node', 'npm', 'Yarn'],
        Utilities: ['Git'],
        Browsers: ['Chrome', 'Edge', 'Firefox', 'Internet Explorer', 'Safari'],
        npmPackages: ['nom-scripts'],
        npmGlobalPackages: ['create-nom-app']
      },
      {
        duplicates: true,
        showNotFound: true
      }
    )
    .then(console.log)
}

export default writeEnvInfo
