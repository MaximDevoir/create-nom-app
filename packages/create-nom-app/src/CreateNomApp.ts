import chalk from 'chalk'
import { sync as spawnSync } from 'cross-spawn'

import path from 'path'
import fs from 'fs-extra'

import logger from './logger'

interface Options {
  projectDirectory: string
  packageManager: {
    manager: 'npm' | 'yarn'
    binary: string
  }
  gitBinary?: string
}

const defaultOptions: Pick<Options, 'gitBinary'> = {
  gitBinary: undefined
}

class CreateNomApp {
  projectName: string
  options: Options

  constructor(projectName: string, options: Options) {
    console.log(`Creating ${chalk.blue(projectName)} app.`)

    // TODO: Do recursive deep-assign
    this.projectName = projectName
    this.options = Object.assign({}, defaultOptions, options)
  }

  ensureProjectDir(): void {
    fs.ensureDirSync(this.options.projectDirectory)
  }

  installPackages(): void {
    console.log(`\nInstalling ${chalk.green('nom-scripts')}. This may take a few minutes...\n`)

    logger.debug('package manager is', this.options.packageManager.manager, 'at', this.options.packageManager.binary)
    const dependencies = ['nom-scripts']
    let installCommand

    // eslint-disable-next-line default-case
    switch (this.options.packageManager.manager) {
      case 'npm':
        installCommand = ['install', '--save']
        break
      case 'yarn':
        installCommand = ['add']
        break
    }

    installCommand.push(...dependencies)
    const install = spawnSync(this.options.packageManager.binary, installCommand, {
      cwd: this.options.projectDirectory,
      stdio: 'inherit'
    })

    if (install.status !== 0) {
      console.error(`Failed to execute ${this.options.packageManager.binary} ${installCommand.join(' ')}`)
      process.exit(1)
    }
  }

  handoff(): void {
    const nomScriptsBin = path.resolve(
      this.options.projectDirectory,
      path.join(this.options.projectDirectory, 'node_modules', '.bin', 'nom-scripts')
    )
    const handoffOptions = JSON.stringify({
      projectName: this.projectName,
      projectDirectory: this.options.projectDirectory,
      packageManager: this.options.packageManager,
      gitBinary: this.options.gitBinary
    })
    const handoffProc = spawnSync(nomScriptsBin, ['init', '--options-string', handoffOptions], {
      stdio: 'inherit',
      cwd: this.options.projectDirectory
    })
    if (handoffProc.status) {
      console.error(
        `Create Nom App failed failed to initialize the project and exited with code ${handoffProc.status}.`
      )
      console.error(handoffProc.error)
      process.exit(handoffProc.status || 1)
    }
  }
}

export default CreateNomApp
