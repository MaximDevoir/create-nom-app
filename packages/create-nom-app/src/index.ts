#!/usr/bin/env node

import './validators/validateNodeVersion'

import chalk from 'chalk'
import commander from 'commander'
import path from 'path'

import logger from './logger'
import writeHelp from './writeHelp'
import writeEnvInfo from './writeEnvInfo'
import validateProjectName from './validators/validateProjectName'
import validateProjectDirectory from './validators/validateProjectDirectory'
import validateNPMVersion from './validators/validateNPMVersion'
import validateYarnVersion from './validators/validateYarnVersion'
import CreateNomApp from './CreateNomApp'
import packageManagers from './package-managers'
import git from './discover-git'

// eslint-disable-next-line @typescript-eslint/no-var-requires
const { version } = require('../package.json')

function main() {
  const cwd = process.cwd()
  let projectDirectory
  let projectName

  const program = new commander.Command('create-nom-app')
    .arguments('<project-name>')
    .usage(`${chalk.blue('<project-name>')} [options]`)
    .action(name => {
      projectName = name
      projectDirectory = path.join(cwd, projectName)
    })
    .on('--help', () => {
      writeHelp()
    })
    .option('--info', 'print environment info')
    .option('--use-npm')
    .option('--use-yarn')
    .option('-v, --version', `output the version number ${version}`, () => {
      console.log(`v${version}`)
      process.exit(0)
    })
    .option('-V', '', () => {
      console.log(`v${version}`)
      process.exit(0)
    })

  program.parse(process.argv)

  if (program.rawArgs.includes('-h') || program.rawArgs.includes('--help')) {
    program.help()
  }

  if (program.info) {
    return writeEnvInfo()
  }

  if (typeof projectName !== 'string') {
    console.log(`${chalk.green('create-nom-app')} ${chalk.blue('<project-name>')} [options]`)
    writeHelp()
    process.exit(0)
  }

  if (program.useNpm && program.useYarn) {
    console.error(chalk.red('Unable to create the project because you have --use-npm and --use-yarn.'))
    console.error(chalk.red('You must remove one of the options.'))
    process.exit(1)
  }

  packageManagers.discoverCommon()

  const preferredPackageManager = ((): false | 'npm' | 'yarn' => {
    if (program.useNpm) {
      return 'npm'
    }

    if (program.useYarn) {
      return 'yarn'
    }

    return false
  })()

  /**
   * If the user has a preference (via --use-{package-manager}) then it will
   * load that manager. If the preferred manager does not exist, an error will
   * be provided and the process will quit.
   *
   * If the user has no preference, then it will search through a list of
   * package managers and return the first existing manager on the system.
   *
   * If no package manager exists on the system, an error is logged and the
   * process quits.
   */
  // eslint-disable-next-line consistent-return
  const usePackageManager = (() => {
    if (preferredPackageManager) {
      if (packageManagers.hasManager(preferredPackageManager)) {
        return preferredPackageManager
      }

      console.error(
        chalk.red(`You preferred the package manager ${preferredPackageManager}, which was not found on your system.`)
      )
      console.error(chalk.red('You can install the package manager you prefer, or remove any preference.'))
      process.exit(1)
    }

    const pkgManagerDefaultPreference = ['yarn', 'npm']

    // eslint-disable-next-line no-restricted-syntax
    for (const manager of pkgManagerDefaultPreference) {
      if (packageManagers.hasManager(manager)) {
        return manager
      }
    }

    console.error(chalk.red('Unable to find any package managers on your system.'))
    console.error(chalk.red(`Searched for ${pkgManagerDefaultPreference.join(', ')}.`))
    console.error(chalk.red('Please install a package manager.'))
    process.exit(1)
  })()

  const packageManagerBinary = packageManagers.getManager(usePackageManager).binary

  logger.debug('pref', preferredPackageManager, ';using', usePackageManager, ';binary', packageManagerBinary)

  if (usePackageManager === 'npm') {
    validateNPMVersion(packageManagers.getManager('npm').version)
  }

  if (usePackageManager === 'yarn') {
    validateYarnVersion(packageManagers.getManager('yarn').version)
  }

  validateProjectName(projectName)
  validateProjectDirectory(projectDirectory)

  git.discoverGit()

  const { gitBinary } = git.gitInfo
  logger.debug('gitInfo', git.gitInfo)

  const app = new CreateNomApp(projectName, {
    projectDirectory,
    packageManager: {
      manager: usePackageManager,
      binary: packageManagerBinary
    },
    gitBinary
  })

  logger.debug('createNomApp', app)

  app.ensureProjectDir()
  app.installPackages()
  app.handoff()

  return undefined
}

main()
