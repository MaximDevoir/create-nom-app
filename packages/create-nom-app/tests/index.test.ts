/* eslint-disable @typescript-eslint/no-var-requires */
describe('main', () => {
  test('blank', () => {
    const spyOnExit = jest.spyOn(process, 'exit').mockImplementation()
    require('../src/index')(process.argv.slice(5))

    expect(spyOnExit).toHaveBeenCalledWith(0)
  })
})
