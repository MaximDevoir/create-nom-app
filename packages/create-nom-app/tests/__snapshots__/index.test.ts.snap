// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Nom App CLI --help, -h display help (--help) 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Usage: create-nom-app [34m<project-name>[39m [options]

Options:
  --info         print environment info
  --use-npm      
  --use-yarn     
  -v, --version  output the version number __mocked_version_0.0.0-development__
  -V             
  -h, --help     output usage information
",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Create Nom App CLI --help, -h display help (-h) 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Usage: create-nom-app [34m<project-name>[39m [options]

Options:
  --info         print environment info
  --use-npm      
  --use-yarn     
  -v, --version  output the version number __mocked_version_0.0.0-development__
  -V             
  -h, --help     output usage information
",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Create Nom App CLI --info display environment information 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "  Create Nom App: __mocked_version_0.0.0-development__",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Create Nom App CLI --info display environment information 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[1m[22m
[1mEnvironment Info:[22m",
    ],
    Array [
      "A mocked call to envinfo.run was made",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Create Nom App CLI --info display environment information 3`] = `[MockFunction]`;

exports[`Create Nom App CLI -v, --version, -V --version 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "v__mocked_version_0.0.0-development__",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Create Nom App CLI -v, --version, -V -V 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "v__mocked_version_0.0.0-development__",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Create Nom App CLI -v, --version, -V -v 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "v__mocked_version_0.0.0-development__",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Create Nom App CLI show usage instructions and exit with code 1 when no project name is supplied 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[32mcreate-nom-app[39m [34m<project-name>[39m [options]",
    ],
    Array [
      "  ",
    ],
    Array [
      "  Only [34m<project-name>[39m is required.",
    ],
    Array [
      "  ",
    ],
    Array [
      "  Example:",
    ],
    Array [
      "    [32mcreate-nom-app[39m [34mmy-first-app[39m",
    ],
    Array [
      "  ",
    ],
    Array [
      "  Run [32mcreate-nom-app[39m --help to see all options.",
    ],
    Array [
      "  ",
    ],
    Array [
      "  If you encounter any issues, feel free to report it to:",
    ],
    Array [
      "   > https://github.com/create-nom-app/create-nom-app/issues",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
